core/templates/fixed_vector.h

```diff
@@ -0,0 +1,157 @@
+	alignas(T) uint8_t _data[CAPACITY * sizeof(T)];
+
+	constexpr static uint32_t DATA_PADDING = MAX(alignof(T), alignof(uint32_t)) - alignof(uint32_t);
+
+public:
+	_FORCE_INLINE_ constexpr FixedVector() = default;
+	constexpr FixedVector(std::initializer_list<T> p_init) {
+		ERR_FAIL_COND(p_init.size() > CAPACITY);
+		for (const T &element : p_init) {
+			memnew_placement(ptr() + _size++, T(element));
+		}
+	}
+
+	template <uint32_t p_capacity, bool p_force_trivial = false>
+	constexpr FixedVector(const FixedVector<T, p_capacity, p_force_trivial> &p_from) {
+		ERR_FAIL_COND(p_from.size() > CAPACITY);
+		if constexpr (std::is_trivially_copyable_v<T>) {
+			// Copy size and the needed elements at once.
+			memcpy((void *)&_size, (void *)&p_from._size, sizeof(_size) + DATA_PADDING + p_from.size() * sizeof(T));
+		} else {
+			for (const T &element : p_from) {
+				memnew_placement(ptr() + _size++, T(element));
+			}
+		}
+	}
+
+	template <uint32_t p_capacity, bool p_force_trivial = false>
+	constexpr FixedVector(FixedVector<T, p_capacity, p_force_trivial> &&p_from) {
+		ERR_FAIL_COND(p_from.size() > CAPACITY);
+		if constexpr (std::is_trivially_copyable_v<T>) {
+			// Copy size and the needed elements at once.
+			memcpy((void *)&_size, (void *)&p_from._size, sizeof(_size) + DATA_PADDING + p_from.size() * sizeof(T));
+		} else {
+			for (T &element : p_from) {
+				memnew_placement(ptr() + _size++, T(std::move(element)));
+			}
+		}
+	}
+
+	~FixedVector() {
+		if constexpr (!std::is_trivially_destructible_v<T>) {
+			for (uint32_t i = 0; i < _size; i++) {
+				ptr()[i].~T();
+			}
+		}
+	}
+
+	_FORCE_INLINE_ constexpr T *ptr() { return (T *)(_data); }
+	_FORCE_INLINE_ constexpr const T *ptr() const { return (const T *)(_data); }
+
+	_FORCE_INLINE_ constexpr operator Span<T>() const { return Span<T>(ptr(), size()); }
+	_FORCE_INLINE_ constexpr Span<T> span() const { return operator Span<T>(); }
+
+	_FORCE_INLINE_ constexpr uint32_t size() const { return _size; }
+	_FORCE_INLINE_ constexpr bool is_empty() const { return !_size; }
+	_FORCE_INLINE_ constexpr bool is_full() const { return _size == CAPACITY; }
+	_FORCE_INLINE_ constexpr uint32_t capacity() const { return CAPACITY; }
+
+	_FORCE_INLINE_ constexpr void clear() { resize(0); }
+
+	template <bool p_ensure_zero = false>
+	constexpr Error resize(uint32_t p_size) {
+		if (!force_trivial && p_size > _size) {
+			ERR_FAIL_COND_V(p_size > CAPACITY, ERR_OUT_OF_MEMORY);
+			memnew_arr_placement<p_ensure_zero>(ptr() + _size, p_size - _size);
+		} else if (p_size < _size) {
+			if constexpr (!std::is_trivially_destructible_v<T>) {
+				for (uint32_t i = p_size; i < _size; i++) {
+					ptr()[i].~T();
+				}
+			}
+		}
+
+		_size = p_size;
+		return OK;
+	}
+
+	constexpr void push_back(const T &p_val) {
+		ERR_FAIL_COND(_size >= CAPACITY);
+		memnew_placement(ptr() + _size, T(p_val));
+		_size++;
+	}
+
+	constexpr void pop_back() {
+		ERR_FAIL_COND(_size == 0);
+		_size--;
+		ptr()[_size].~T();
```

Comment:
no `if constexpr (!std::is_trivially_destructible_v<T>)` here ?

Reply:
I think it just gets compiled away for trivial types:

https://godbolt.org/z/71oWoY19P

I think technically we should expect the same in all the loop cases as well, but the reason I'm checking there is to make it easier for the compiler / not to risk it leaving in the loop for whatever reason.